### PR TITLE
added JWK support

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: gateway/utils/jwt/jwt.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		for t := range tick.C {
+			for kid, secret := range j.jwkSecrets {
+				if secret.refreshTime.After(t) {
+					jwkSecretInfo, err := getJWKRefreshTime(secret.url)
+					if err != nil {
+						_ = helpers.Logger.LogError("", fmt.Sprintf("Unable to refresh jwk keys having kid (%s) and url (%s)", kid, secret.url), err, nil)
+					}
+					j.jwkSecrets[kid] = jwkSecretInfo
+				}
+			}
```
Comment (YourTechBud) [inline gateway/utils/jwt/jwt.go:43]: put this in function and acquire write lock before the loop

File: gateway/utils/jwt/jwt.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+const defaultRefreshTime = 1 * time.Hour
+
+// New initializes the package
+func New() *JWT {
+	j := new(JWT)
+	go func() {
+		tick := time.NewTicker(defaultRefreshTime)
+		for t := range tick.C {
+			for kid, secret := range j.jwkSecrets {
+				if secret.refreshTime.After(t) {
```
Comment (YourTechBud) [inline gateway/utils/jwt/jwt.go:36]:
```suggestion
				if secret.refreshTime.Before(t) {
```

File: gateway/utils/jwt/jwt.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+					jwkSecretInfo, err := getJWKRefreshTime(secret.url)
+					if err != nil {
+						_ = helpers.Logger.LogError("", fmt.Sprintf("Unable to refresh jwk keys having kid (%s) and url (%s)", kid, secret.url), err, nil)
```
Comment (YourTechBud) [inline gateway/utils/jwt/jwt.go:39]: Add `continue`

File: gateway/utils/jwt/jwt.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+				}
+			}
+		}
+	}()
+	return j
+}
+
+// SetSecrets set the internal fields of jwt struct
+func (j *JWT) SetSecrets(secrets []*config.Secret) error {
+	newJwkSecrets := map[string]*jwkSecret{}
```
Comment (YourTechBud) [inline gateway/utils/jwt/jwt.go:51]: Acquire write lock first

File: gateway/utils/jwt/jwt.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	url         string
+	refreshTime time.Time
+	set         *jwk.Set
+}
+
+const defaultRefreshTime = 1 * time.Hour
+
+// New initializes the package
+func New() *JWT {
+	j := new(JWT)
```
Comment (YourTechBud) [inline gateway/utils/jwt/jwt.go:31]: I think you should initialise each map inside `JWT` as well.

File: gateway/utils/jwt/operations.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/dgrijalva/jwt-go"
+	"github.com/spaceuptech/helpers"
+
+	"github.com/spaceuptech/space-cloud/gateway/config"
+	"github.com/spaceuptech/space-cloud/gateway/model"
+)
+
+// ParseToken verifies the token
+func (j *JWT) ParseToken(ctx context.Context, token string) (map[string]interface{}, error) {
+	parser := jwt.Parser{}
```
Comment (YourTechBud) [inline gateway/utils/jwt/operations.go:18]: Acquire read lock first

File: gateway/utils/jwt/operations.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		}
+		// check if kid belongs to a normal token with kid header
+		obj, ok := j.staticSecrets[kid.(string)]
+		if ok {
+			if obj.Alg == config.RS256Public {
+				obj.Alg = config.RS256
+			}
+			return j.verifyTokenSignature(ctx, token, obj)
+		}
+		return nil, helpers.Logger.LogError(helpers.GetRequestID(ctx), "", nil, nil)
```
Comment (YourTechBud) [inline gateway/utils/jwt/operations.go:39]:
```suggestion
		return nil, helpers.Logger.LogError(helpers.GetRequestID(ctx), "No secret with given kid found", nil, map[string]interface{}{"kid": kid})
```

File: gateway/utils/jwt/operations.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+		if err == nil {
+			return claims, nil
+		}
+	}
+	return nil, helpers.Logger.LogError(helpers.GetRequestID(ctx), "Authentication secrets not set in space cloud config", err, nil)
+}
+
+// CreateToken create a token with primary secret
+func (j *JWT) CreateToken(ctx context.Context, tokenClaims model.TokenClaims) (string, error) {
+	claims := jwt.MapClaims{}
```
Comment (YourTechBud) [inline gateway/utils/jwt/operations.go:61]: Acquire read lock

File: gateway/utils/jwt/operations.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	}
+	var tokenString string
+	var err error
+	// Add expiry of one week
+	claims["exp"] = time.Now().Add(24 * 7 * time.Hour).Unix()
+	for _, s := range j.staticSecrets {
+		if s.IsPrimary {
+			switch s.Alg {
+			case config.RS256:
+				token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
```
Comment (YourTechBud) [inline gateway/utils/jwt/operations.go:73]: inject the `kid` in `token.Header`

File: gateway/utils/jwt/operations.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+				if err != nil {
+					return "", err
+				}
+				tokenString, err = token.SignedString(signKey)
+				if err != nil {
+					return "", err
+				}
+				return tokenString, nil
+			case config.HS256, "":
+				token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
```
Comment (YourTechBud) [inline gateway/utils/jwt/operations.go:84]: inject the `kid` in `token.Header`

File: runner/model/service.go [hunk 11]
```diff
@@ -88,6 +88,7 @@ const (
 
 // Affinity describes the affinity rules of a service
 type Affinity struct {
+	ID               string             `json:"id" yaml:"id"`
```
Comment (YourTechBud) [inline runner/model/service.go:91]: What is this for?

**Review by YourTechBud — CHANGES_REQUESTED** [review]

File: runner/model/service.go (hunk 11 above)
Comment (sharadregoti) [reply runner/model/service.go:91]: @HeyItsJs said he wanted something to uniquely identify an affinity object

Comment (YourTechBud) [reply runner/model/service.go:91]: Shouldn't you store it somewhere then? Would become a problem while getting the affinities in k8s

Comment (sharadregoti) [reply runner/model/service.go:91]: I am not storing it anywhere, while fetching the service, i'll generate a new kid

File: gateway/utils/jwt/operations.go [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+	var tokenString string
+	var err error
+	// Add expiry of one week
+	claims["exp"] = time.Now().Add(24 * 7 * time.Hour).Unix()
+	for _, s := range j.staticSecrets {
+		if s.IsPrimary {
+			switch s.Alg {
+			case config.RS256:
+				token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+				token.Header = map[string]interface{}{"alg": config.RS256, "kid": s.KID}
```
Comment (YourTechBud) [inline gateway/utils/jwt/operations.go:80]:
```suggestion
				token.Header["kid"] = s.KID
```

File: gateway/utils/jwt/operations.go [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+					return "", err
+				}
+				tokenString, err = token.SignedString(signKey)
+				if err != nil {
+					return "", err
+				}
+				return tokenString, nil
+			case config.HS256, "":
+				token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+				token.Header = map[string]interface{}{"alg": config.HS256, "kid": s.KID}
```
Comment (YourTechBud) [inline gateway/utils/jwt/operations.go:92]:
```suggestion
				token.Header["kid"] = s.KID
```

**Review by YourTechBud — CHANGES_REQUESTED** [review]

**Review by YourTechBud — APPROVED** [review]

**Pull request merged**